src/python/WMCore/MicroService/Unified/MSOutputTemplate.py

```diff
@@ -0,0 +1,228 @@
+    "transferStatus": "blah", # either "pending" or "done",
+    "transferIDs": ["list of transfer IDs"],
+    "numberOfCopies": integer
+    }
+    """
+    # TODO:
+    #    To add an identifier parameter to __init__
+    #    to distinguish inbound from outbound documents.
+
+    # DONE:
+    #    to figure out how to pass params as dict not only kwargs so I can use
+    #    the same logic for both inbound and outbound documents
+
+    # TODO:
+    #    to implement document mutex in a model similar to the one described here:
+    #    https://stackoverflow.com/questions/32728670/mutex-with-mongodb
+    #    ```
+    #    db.collection.update({done: false, taken: false},{$set: {taken: true, takenBy: myIp}});
+    #    ```
+
+    # TODO:
+    #    as soon as any change in the instantiated object happens ( any call of
+    #    a set method) the same change must be updated in Mongodb too -
+    #    this may happen from outside or from the object itself. The former method
+    #    keeps the document independent from the underlying Database technology
+
+    def __init__(self, doc=None, **kwargs):
+        super(MSOutputTemplate, self).__init__(**kwargs)
+        # super(MSOutputTemplate, self).__init__()
+
+        docTemplate = [
+            ('_id', None, unicode),
+            ('RequestName', None, unicode),
+            ('RequestStatus', None, unicode),
+            ('creationTime', None, Timestamp),
+            ('lastUpdate', None, float),
+            ('isRelVal', None, bool),
+            ('isTaken', None, bool),
+            ('OutputDatasets', None, list),
+            ('destination', None, list),
+            ('destinationOutputMap', None, dict),
+            ('campaignOutputMap', None, dict),
+            ('transferStatus', None, unicode),
+            ('transferIDs', None, unicode),
+            ('numberOfCopies', None, int)]
+
+        self.docTemplate = docTemplate
+        self.required = ['_id', 'RequestName', 'creationTime']
+        self.allowEmpty = ['OutputDatasets']
+
+        myDoc = dict()
+        for tup in docTemplate:
+            myDoc.update({tup[0]: tup[1]})
+
+        # set creation time - if already present in the document it will be
+        # overwritten with the value from the document itself during the _checkAttr
+        # call and this value here will be ignored
+        myDoc['creationTime'] = Timestamp(int(time.time()), 1)
+
+        # if no document was passed consider only **kwargs
+        if doc is not None:
+            # Search the keys from the document template for their equivalent into
+            # the document passed and fill them into internal document so that they
+            # can pass the needed checks later, throw the unneeded/unrecognised
+            # key/values from the passed document
+            for key in map(lambda tup: tup[0], docTemplate):
+                if key in doc.keys():
+                    myDoc[key] = deepcopy(doc[key])
```

Comment:
If we will operate on the mongo doc, any reason not to pass the value by reference?

Reply:
Yes, one  reason and it is quite solid - because of that line here [1]. I am eager to drop everything that I do not need from that huge document that comes  from Reqmgr. And with that line there things will vanish from the other document too ,which at that moment is not uploaded to MongoDB yet.

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSOutput.py#L346